File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/data/ServiceCatalog.js
/**
 * Copyright (c) 2020, WSO2 Inc. (http://wso2.com) All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import cloneDeep from 'lodash.clonedeep';
import Utils from './Utils';
import APIClientFactory from './APIClientFactory';

/**
  * An abstract representation of a Service Catalog
  */
class ServiceCatalog {
    constructor(kwargs) {
        this.client = new APIClientFactory().getAPIClient(Utils.getCurrentEnvironment(),
            Utils.CONST.SERVICE_CATALOG_CLIENT).client;
        const properties = kwargs;
        Utils.deepFreeze(properties);
        this._data = properties;
        for (const key in properties) {
            if (Object.prototype.hasOwnProperty.call(properties, key)) {
                this[key] = properties[key];
            }
        }
    }

    /**
     * @param data
     * @returns {object} Metadata for API request
     */
    static _requestMetaData(data = {}) {
        return {
            requestContentType: data['Content-Type'] || 'application/json',
        };
    }

    /**
     *
     * Instance method of the ServiceCatalog class to provide raw JSON object
     * which is Service body friendly to use with REST api requests
     * Use this method instead of accessing the private _data object for
     * converting to a JSON representation of an API object.
     * Note: This is deep coping, Use sparingly, Else will have a bad impact on performance
     * Basically this is the revers operation in constructor.
     * This method simply iterate through all the object properties (excluding the properties in `excludes` list)
     * and copy their values to new object.
     * So use this method with care!!
     * @memberof API
     * @param {Array} [userExcludes=[]] List of properties that are need to be excluded from the generated JSON object
     * @returns {JSON} JSON representation of the API
     */
    toJSON(userExcludes = []) {
        const copy = {};
        const excludes = [...userExcludes];
        for (const prop in this) {
            if (!excludes.includes(prop)) {
                copy[prop] = cloneDeep(this[prop]);
            }
        }
        return copy;
    }

    /**
     * Add sample service
     * @returns {promise} Add sample promise.
     */
    static addService(serviceMetadata, inlineContent) {
        const serviceCatalog = new APIClientFactory()
            .getAPIClient(Utils.getCurrentEnvironment(), Utils.CONST.SERVICE_CATALOG_CLIENT)
            .client;
        const definitionFile = new File([JSON.stringify(inlineContent)],
            'definitionFile.json', { type: 'application/json', lastModified: new Date().getTime() });
        const serviceMetadataFile = new File([JSON.stringify(serviceMetadata)],
            'serviceMetadata.json', { type: 'application/json', lastModified: new Date().getTime() });
        const requestBody = {
            serviceMetadata: serviceMetadataFile,
            definitionFile,
        };
        const promisedCatalogSampleService = serviceCatalog.then((client) => {
            return client.apis.Services
                .addService({}, {
                    requestBody,
                }, this._requestMetaData({
                    'Content-Type': 'multipart/form-data',
                }));
        });
        return promisedCatalogSampleService.then((response) => response.body);
    }

    /**
     * Get Settings
     * @returns {promise} Settings promise.
     */
    static getSettings() {
        const serviceCatalog = new APIClientFactory()
            .getAPIClient(Utils.getCurrentEnvironment(), Utils.CONST.SERVICE_CATALOG_CLIENT)
            .client;
        const promisedServiceCatalogSettings = serviceCatalog.then((client) => {
            return client.apis.Services.getSettings();
        });
        return promisedServiceCatalogSettings.then((response) => response.body);
    }

    /**
     * Get details of Services
     * @returns {promise} Service Entry promise.
     */
    static searchServices(limit = 10, offset = 0) {
        const serviceCatalog = new APIClientFactory()
            .getAPIClient(Utils.getCurrentEnvironment(), Utils.CONST.SERVICE_CATALOG_CLIENT)
            .client;
        return serviceCatalog.then((client) => {
            return client.apis.Services.searchServices({ limit, offset });
        });
    }

    /**
     * Get details of Service by key
     * @returns {promise} Service Entry promise.
     */
    static searchServiceByKey(key) {
        const serviceCatalog = new APIClientFactory()
            .getAPIClient(Utils.getCurrentEnvironment(), Utils.CONST.SERVICE_CATALOG_CLIENT)
            .client;
        return serviceCatalog.then((client) => {
            return client.apis.Services.searchServices(
                {
                    key,
                },
                this._requestMetaData(),
            );
        });
    }

    /**
     * Delete a Service
     * @param id {string} UUID of the service.
     * @returns {promise} Promise.
     */
    static deleteService(id) {
        const serviceCatalog = new APIClientFactory()
            .getAPIClient(Utils.getCurrentEnvironment(), Utils.CONST.SERVICE_CATALOG_CLIENT)
            .client;
        return serviceCatalog.then((client) => {
            return client.apis.Services.deleteService({ serviceId: id });
        }).then((response) => response.body);
    }

    /**
     * Get a service by ID
     * @param id {string} UUID of the service.
     * @returns {promise} Service Entry promise.
     */
    static getServiceById(id) {
        const serviceCatalog = new APIClientFactory()
            .getAPIClient(Utils.getCurrentEnvironment(), Utils.CONST.SERVICE_CATALOG_CLIENT)
            .client;
        const promisedServices = serviceCatalog.then((client) => {
            return client.apis.Services.getServiceById(
                {
                    serviceId: id,
                },
                this._requestMetaData(),
            );
        });
        return promisedServices.then((response) => response.body);
    }

    /**
     * Get a service by Name
     * @param id {string} name of the service.
     * @returns {promise} Service Entry promise.
     */
    static getServiceByName(info) {
        const serviceCatalog = new APIClientFactory()
            .getAPIClient(Utils.getCurrentEnvironment(), Utils.CONST.SERVICE_CATALOG_CLIENT)
            .client;
        const promisedServices = serviceCatalog.then((client) => {
            return client.apis.Services.searchServices(
                {
                    name: info.name,
                },
                this._requestMetaData(),
            );
        });
        return promisedServices.then((response) => response.body);
    }

    /**
     * Get the definition of a service
     * @param id {string} UUID of the service.
     * @returns {promise} Service Entry promise.
     */
    static getServiceDefinition(id) {
        const serviceCatalog = new APIClientFactory()
            .getAPIClient(Utils.getCurrentEnvironment(), Utils.CONST.SERVICE_CATALOG_CLIENT)
            .client;
        const promisedServices = serviceCatalog.then((client) => {
            return client.apis.Services.getServiceDefinition(
                {
                    serviceId: id,
                },
                this._requestMetaData(),
            );
        });
        return promisedServices.then((response) => response.body);
    }

    /**
     * Get the usages of a service
     * @param id {string} UUID of the service.
     * @returns {promise} Service Entry API promise.
     */
    static getAPIUsages(id) {
        const serviceCatalog = new APIClientFactory()
            .getAPIClient(Utils.getCurrentEnvironment(), Utils.CONST.SERVICE_CATALOG_CLIENT)
            .client;
        const promisedServices = serviceCatalog.then((client) => {
            return client.apis.Services.getServiceUsage(
                {
                    serviceId: id,
                },
                this._requestMetaData(),
            );
        });
        return promisedServices.then((response) => response.body);
    }
}

export default ServiceCatalog;

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/ServiceCatalog.java
/*
 *  Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
 *
 *  WSO2 Inc. licenses this file to you under the Apache License,
 *  Version 2.0 (the "License"); you may not use this file except
 *  in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.wso2.carbon.apimgt.api;

import org.wso2.carbon.apimgt.api.model.API;
import org.wso2.carbon.apimgt.api.model.ServiceEntry;
import org.wso2.carbon.apimgt.api.model.ServiceFilterParams;

import java.util.List;

public interface ServiceCatalog {

    /**
     * Adds a new Service Catalog
     *
     * @param serviceEntry ServiceCatalogInfo
     * @param tenantId     Tenant Identifier
     * @param user         Logged in user name
     * @return ServiceCatalogId UUID of the created Service Catalog ID
     * @throws APIManagementException if failed to add ServiceCatalogInfo
     */
    String addService(ServiceEntry serviceEntry, int tenantId, String user) throws APIManagementException;

    /**
     * Imports the List of Services to Service Catalog
     * @param serviceList
     * @param tenantId
     * @param username
     * @throws APIManagementException
     */
    List<ServiceEntry> importServices(List<ServiceEntry> serviceList, int tenantId, String username, boolean overwrite)
            throws APIManagementException;

    /**
     * Update an existing Service Catalog
     *
     * @param serviceEntry ServiceCatalogInfo
     * @param tenantId     Tenant Identifier
     * @param user         Logged in user name
     * @throws APIManagementException if failed to add ServiceCatalogInfo
     */
    void updateService(ServiceEntry serviceEntry, int tenantId, String user) throws APIManagementException;

    /**
     * Returns details of an Service Catalog
     *
     * @param serviceCatalogId ServiceCatalog Identifier
     * @param tenantId         Tenant Identifier
     * @return An ServiceCatalogInfo object related to the given identifier or null
     * @throws APIManagementException if failed to get details of an Service Catalog
     */
    ServiceEntry getServiceByUUID(String serviceCatalogId, int tenantId)
            throws APIManagementException;

    /**
     * Deletes a Service from Catalog
     *
     * @param serviceKey Service key
     *
     * @throws APIManagementException if failed to delete the Service Catalog
     */
    void deleteService(String serviceKey, int tenantId) throws APIManagementException;

    /**
     * Returns details of all Service Catalogs belong to a given tenant
     *
     * @param tenantId Tenant Identifier
     * @return A list of ServiceCatalogInfo objects
     * @throws APIManagementException if failed to get details of Service Catalogs
     */
    List<ServiceEntry> getService(int tenantId) throws APIManagementException;

    /**
     * Adds a new end-point definition
     *
     * @param serviceEntry End-point related information from ServiceCatalogEntry
     * @return ServiceCatalogId UUID of the created Service Catalog ID
     * @throws APIManagementException if failed to add ServiceCatalogInfo
     */
    String addEndPointDefinition(ServiceEntry serviceEntry, String uuid) throws APIManagementException;

    /**
     * Get MD5 hash value of a service endpoint
     *
     * @param serviceEntry EndPoint related information
     * @return ServiceCatalogInfo Endpoint information with md5 hash value
     * @throws APIManagementException if failed to add ServiceCatalogInfo
     */
    ServiceEntry getMD5Hash(ServiceEntry serviceEntry, int tenantId) throws APIManagementException;

    /**
     * Get MD5 hash value of a service endpoint
     *
     * @param key Service key unique to each tenant
     * @return Service md5 hash value
     * @throws APIManagementException if failed to add ServiceCatalogInfo
     */
    String getMD5HashByKey(String key, int tenantId) throws APIManagementException;

    /**
     * Get information of a service endpoint by key
     *
     * @param key Service key unique to each tenant
     * @param tenantId Tenant Identifier
     * @return ServiceEntry object including endpoint information
     * @throws APIManagementException if failed to add ServiceCatalogInfo
     */
    ServiceEntry getServiceByKey(String key, int tenantId) throws APIManagementException;

    /**
     * Get metadata and endpoint definition resources of a service endpoint
     *
     * @param name Service name
     * @param version Version of service
     * @param tenantId Tenant Identifier
     * @return ServiceEntry Endpoint resources
     * @throws APIManagementException if failed to add ServiceCatalogInfo
     */
    ServiceEntry getEndPointResourcesByNameAndVersion(String name, String version, int tenantId)
            throws APIManagementException;

    /**
     * Search Services in Service Catalog
     * @param filterParams
     * @param tenantId
     * @param shrink
     * @return
     * @throws APIManagementException
     */
    List<ServiceEntry> getServices(ServiceFilterParams filterParams, int tenantId, boolean shrink)
            throws APIManagementException;

    /**
     * Retrieve the Info of APIs that use the Service
     *
     * @param serviceId UUID of the service
     * @param tenantId Logged-In user's tenant Id
     * @return List of APIs
     * @throws APIManagementException
     */
    List<API> getServiceUsage(String serviceId, int tenantId) throws APIManagementException;
}
